#### Code by Larissa Evaldt

#### Importing the libraries

In [1]:
import glob, os
import pandas as pd

#### After running the Initial Exploration Jupyter notebook in the years 2017, 18 and 19 we had 3 files so we need to combine those 3 files into one large dataset. So let's read the 3 csv files of years 2017, 2018 and 2019 into one dataframe

In [2]:
df = pd.concat(map(pd.read_csv, glob.glob(os.path.join('', "*.csv"))))

#### Checking if all it worked. 2017 file had 17,953 rows, 2018 had 18,625 and 2019 had 9,881

In [3]:
df.shape

(46458, 18)

#### We can see that it worked because the number of rows in the dataframe equals the sum of all 3 files. Let's have a look in the top 5 rows:

In [4]:
df.head()

Unnamed: 0 Date of Sale (dd/mm/yyyy)  \
0           0                02/01/2018   
1           1                02/01/2018   
2           2                02/01/2018   
3           3                02/01/2018   
4           4                02/01/2018   

                                    Address Postal Code  County  Price (�)  \
0                 1 ABBEY ST, HOWTH, DUBLIN    Dublin 7  Dublin   710000.0   
1         1 ROSEVILLE, LOWER ROAD, SHANKILL   Dublin 18  Dublin   627230.7   
2            19 ULVERTON RD, DALKEY, DUBLIN         NaN  Dublin  1130000.0   
3        236 PHIBSBORO RD, DUBLIN 7, DUBLIN    Dublin 7  Dublin   277000.0   
4  3 COLLEGE COURT, PORTRANE ROAD, DONABATE         NaN  Dublin  348017.62   

  Not Full Market Price VAT Exclusive                Description of Property  \
0                    No            No  Second-Hand Dwelling house /Apartment   
1                    No            No  Second-Hand Dwelling house /Apartment   
2                    No            No  Second-Hand Dwelling house /Apartment   
3                    No            No  Second-Hand Dwelling house /Apartment   
4                    No            No  Second-Hand Dwelling house /Apartment   

  Property Size Description        long        lat street_number  \
0                       NaN  -6.0646082  53.388117             1   
1                       NaN  -6.1253428  53.232940             1   
2                       NaN   -6.107343  53.279602            19   
3                       NaN  -6.2733223  53.355613           236   
4                       NaN  -6.1412172  53.489876           NaN   

               route      neighborhood postal_town         county postal_code  
0       Abbey Street               NaN         NaN  County Dublin    D13 R2P6  
1          Roseville               NaN         NaN  County Dublin    D18 YY71  
2      Ulverton Road            Dalkey         NaN  County Dublin    A96 HY04  
3  Phibsborough Road               NaN    Dublin 7  County Dublin    D07 N2E0  
4      Portrane Road  Ballalease North         NaN  County Dublin         NaN

#### There is an extra index column that is not need so I am going to delete it

In [5]:
df = df.drop(columns=['Unnamed: 0'])

In [6]:
df.head()

Date of Sale (dd/mm/yyyy)                                   Address  \
0                02/01/2018                 1 ABBEY ST, HOWTH, DUBLIN   
1                02/01/2018         1 ROSEVILLE, LOWER ROAD, SHANKILL   
2                02/01/2018            19 ULVERTON RD, DALKEY, DUBLIN   
3                02/01/2018        236 PHIBSBORO RD, DUBLIN 7, DUBLIN   
4                02/01/2018  3 COLLEGE COURT, PORTRANE ROAD, DONABATE   

  Postal Code  County  Price (�) Not Full Market Price VAT Exclusive  \
0    Dublin 7  Dublin   710000.0                    No            No   
1   Dublin 18  Dublin   627230.7                    No            No   
2         NaN  Dublin  1130000.0                    No            No   
3    Dublin 7  Dublin   277000.0                    No            No   
4         NaN  Dublin  348017.62                    No            No   

                 Description of Property Property Size Description  \
0  Second-Hand Dwelling house /Apartment                       NaN   
1  Second-Hand Dwelling house /Apartment                       NaN   
2  Second-Hand Dwelling house /Apartment                       NaN   
3  Second-Hand Dwelling house /Apartment                       NaN   
4  Second-Hand Dwelling house /Apartment                       NaN   

         long        lat street_number              route      neighborhood  \
0  -6.0646082  53.388117             1       Abbey Street               NaN   
1  -6.1253428  53.232940             1          Roseville               NaN   
2   -6.107343  53.279602            19      Ulverton Road            Dalkey   
3  -6.2733223  53.355613           236  Phibsborough Road               NaN   
4  -6.1412172  53.489876           NaN      Portrane Road  Ballalease North   

  postal_town         county postal_code  
0         NaN  County Dublin    D13 R2P6  
1         NaN  County Dublin    D18 YY71  
2         NaN  County Dublin    A96 HY04  
3    Dublin 7  County Dublin    D07 N2E0  
4         NaN  County Dublin         NaN

#### Writing to csv

In [7]:
df.to_csv('df_allyears_coords.csv')

#### Reading the csv created above with all the years

In [8]:
df = pd.read_csv('df_allyears_coords.csv', index_col=[0])

/Users/larissaevaldt/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [9]:
df.shape

(46458, 17)

In [10]:
df.head()

Date of Sale (dd/mm/yyyy)                                   Address  \
0                02/01/2018                 1 ABBEY ST, HOWTH, DUBLIN   
1                02/01/2018         1 ROSEVILLE, LOWER ROAD, SHANKILL   
2                02/01/2018            19 ULVERTON RD, DALKEY, DUBLIN   
3                02/01/2018        236 PHIBSBORO RD, DUBLIN 7, DUBLIN   
4                02/01/2018  3 COLLEGE COURT, PORTRANE ROAD, DONABATE   

  Postal Code  County  Price (�) Not Full Market Price VAT Exclusive  \
0    Dublin 7  Dublin   710000.0                    No            No   
1   Dublin 18  Dublin   627230.7                    No            No   
2         NaN  Dublin  1130000.0                    No            No   
3    Dublin 7  Dublin   277000.0                    No            No   
4         NaN  Dublin  348017.62                    No            No   

                 Description of Property Property Size Description  \
0  Second-Hand Dwelling house /Apartment                       NaN   
1  Second-Hand Dwelling house /Apartment                       NaN   
2  Second-Hand Dwelling house /Apartment                       NaN   
3  Second-Hand Dwelling house /Apartment                       NaN   
4  Second-Hand Dwelling house /Apartment                       NaN   

         long        lat street_number              route      neighborhood  \
0  -6.0646082  53.388117             1       Abbey Street               NaN   
1  -6.1253428  53.232940             1          Roseville               NaN   
2   -6.107343  53.279602            19      Ulverton Road            Dalkey   
3  -6.2733223  53.355613           236  Phibsborough Road               NaN   
4  -6.1412172  53.489876           NaN      Portrane Road  Ballalease North   

  postal_town         county postal_code  
0         NaN  County Dublin    D13 R2P6  
1         NaN  County Dublin    D18 YY71  
2         NaN  County Dublin    A96 HY04  
3    Dublin 7  County Dublin    D07 N2E0  
4         NaN  County Dublin         NaN

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46458 entries, 0 to 17951
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Date of Sale (dd/mm/yyyy)  46458 non-null  object 
 1   Address                    46458 non-null  object 
 2   Postal Code                29767 non-null  object 
 3   County                     46458 non-null  object 
 4   Price (�)                  46458 non-null  object 
 5   Not Full Market Price      46458 non-null  object 
 6   VAT Exclusive              46458 non-null  object 
 7   Description of Property    46458 non-null  object 
 8   Property Size Description  8294 non-null   object 
 9   long                       46404 non-null  object 
 10  lat                        46404 non-null  float64
 11  street_number              33257 non-null  object 
 12  route                      43053 non-null  object 
 13  neighborhood               32648 non-null  obj

#### We can see that date is of type String so let's convert it to DateTime format

In [12]:
df['Date of Sale (dd/mm/yyyy)'] = pd.to_datetime(df['Date of Sale (dd/mm/yyyy)'], format='%d/%m/%Y')

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46458 entries, 0 to 17951
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Date of Sale (dd/mm/yyyy)  46458 non-null  datetime64[ns]
 1   Address                    46458 non-null  object        
 2   Postal Code                29767 non-null  object        
 3   County                     46458 non-null  object        
 4   Price (�)                  46458 non-null  object        
 5   Not Full Market Price      46458 non-null  object        
 6   VAT Exclusive              46458 non-null  object        
 7   Description of Property    46458 non-null  object        
 8   Property Size Description  8294 non-null   object        
 9   long                       46404 non-null  object        
 10  lat                        46404 non-null  float64       
 11  street_number              33257 non-null  object        
 12  rout

#### By checking the first 5 and last 5 rows I noticed that it was not sorted by year, it was starting in 2018 then 2019 and 2017

In [14]:
df.head()

Date of Sale (dd/mm/yyyy)                                   Address  \
0                2018-01-02                 1 ABBEY ST, HOWTH, DUBLIN   
1                2018-01-02         1 ROSEVILLE, LOWER ROAD, SHANKILL   
2                2018-01-02            19 ULVERTON RD, DALKEY, DUBLIN   
3                2018-01-02        236 PHIBSBORO RD, DUBLIN 7, DUBLIN   
4                2018-01-02  3 COLLEGE COURT, PORTRANE ROAD, DONABATE   

  Postal Code  County  Price (�) Not Full Market Price VAT Exclusive  \
0    Dublin 7  Dublin   710000.0                    No            No   
1   Dublin 18  Dublin   627230.7                    No            No   
2         NaN  Dublin  1130000.0                    No            No   
3    Dublin 7  Dublin   277000.0                    No            No   
4         NaN  Dublin  348017.62                    No            No   

                 Description of Property Property Size Description  \
0  Second-Hand Dwelling house /Apartment                       NaN   
1  Second-Hand Dwelling house /Apartment                       NaN   
2  Second-Hand Dwelling house /Apartment                       NaN   
3  Second-Hand Dwelling house /Apartment                       NaN   
4  Second-Hand Dwelling house /Apartment                       NaN   

         long        lat street_number              route      neighborhood  \
0  -6.0646082  53.388117             1       Abbey Street               NaN   
1  -6.1253428  53.232940             1          Roseville               NaN   
2   -6.107343  53.279602            19      Ulverton Road            Dalkey   
3  -6.2733223  53.355613           236  Phibsborough Road               NaN   
4  -6.1412172  53.489876           NaN      Portrane Road  Ballalease North   

  postal_town         county postal_code  
0         NaN  County Dublin    D13 R2P6  
1         NaN  County Dublin    D18 YY71  
2         NaN  County Dublin    A96 HY04  
3    Dublin 7  County Dublin    D07 N2E0  
4         NaN  County Dublin         NaN

In [15]:
df.tail()

Date of Sale (dd/mm/yyyy)                                   Address  \
17947                2017-12-29  13 Boyd Avenue, Honeypark, Dun Laoghaire   
17948                2017-12-29    14 ST ANTHONY'S ROAD, RIALTO, DUBLIN 8   
17949                2017-12-29        8 ST ANTHONYS RD, RIALTO, DUBLIN 8   
17950                2017-12-30     39 GLENCARRIG DR, FIRHOUSE, DUBLIN 24   
17951                2017-12-30  98 KILDONAN AVE, FINGLAS WEST, DUBLIN 11   

      Postal Code  County    Price (�) Not Full Market Price VAT Exclusive  \
17947         NaN  Dublin  �640,968.00                    No           Yes   
17948         NaN  Dublin  �215,000.00                    No            No   
17949    Dublin 8  Dublin  �215,000.00                    No            No   
17950   Dublin 24  Dublin  �350,000.00                    No            No   
17951   Dublin 11  Dublin  �220,000.00                    No            No   

                     Description of Property  \
17947          New Dwelling house /Apartment   
17948  Second-Hand Dwelling house /Apartment   
17949  Second-Hand Dwelling house /Apartment   
17950  Second-Hand Dwelling house /Apartment   
17951  Second-Hand Dwelling house /Apartment   

                               Property Size Description      long        lat  \
17947  greater than or equal to 38 sq metres and less... -6.142628  53.279831   
17948                                                NaN -6.294655  53.336367   
17949                                                NaN -6.294849  53.336090   
17950                                                NaN -6.349446  53.275307   
17951                                                NaN -6.311186  53.392124   

      street_number                 route neighborhood postal_town  \
17947            13           Boyd Avenue    Honeypark         NaN   
17948            14  Saint Anthony's Road          NaN    Dublin 8   
17949             8  Saint Anthony's Road          NaN    Dublin 8   
17950            39      Glencarrig Drive     Tallaght   Dublin 24   
17951            98       Kildonan Avenue          NaN   Dublin 11   

              county postal_code  
17947  County Dublin    A96 H0C2  
17948  County Dublin    D08 H3V2  
17949  County Dublin    D08 V0V5  
17950  County Dublin    D24 W6VX  
17951  County Dublin    D11 P5V6

#### Sorting dataframe by date

In [16]:
df = df.sort_values(by='Date of Sale (dd/mm/yyyy)')

In [17]:
df

Date of Sale (dd/mm/yyyy)  \
1                   2017-01-01   
0                   2017-01-01   
13                  2017-01-01   
12                  2017-01-01   
11                  2017-01-01   
...                        ...   
9876                2019-08-16   
9877                2019-08-16   
9878                2019-08-16   
9879                2019-08-16   
9873                2019-08-16   

                                                Address Postal Code  County  \
1     11 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST    Dublin 3  Dublin   
0      1 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST    Dublin 3  Dublin   
13    9 TEAL HOUSE, REDCOURT OAKS SEAFIELD RD EAST, ...    Dublin 3  Dublin   
12    8 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EA...    Dublin 3  Dublin   
11                 5 SEAFIELD GROVE, CLONTARF, DUBLIN 3    Dublin 3  Dublin   
...                                                 ...         ...     ...   
9876           73 CLANMAURICE RD, DONNYCARNEY, DUBLIN 5    Dublin 5  Dublin   
9877              8 ALEXANDRA COURT, DUNDRUM, DUBLIN 16   Dublin 14  Dublin   
9878          APT 1, 3 DERMOT ST, CLONGRIFFIN DUBLIN 13   Dublin 13  Dublin   
9879  APT 416, THE CUBES 8, BEACON SOUTH QUARTER DUB...   Dublin 18  Dublin   
9873         20 FINGLASWOOD RD, FINGLAS WEST, DUBLIN 11   Dublin 11  Dublin   

        Price (�) Not Full Market Price VAT Exclusive  \
1     �242,424.00                   Yes            No   
0     �242,424.00                   Yes            No   
13     �66,000.00                   Yes            No   
12    �271,800.00                   Yes            No   
11    �830,000.00                    No            No   
...           ...                   ...           ...   
9876     320000.0                    No            No   
9877     285000.0                    No            No   
9878     215000.0                    No            No   
9879     365000.0                    No            No   
9873     290000.0                    No            No   

                    Description of Property Property Size Description  \
1     Second-Hand Dwelling house /Apartment                       NaN   
0     Second-Hand Dwelling house /Apartment                       NaN   
13    Second-Hand Dwelling house /Apartment                       NaN   
12    Second-Hand Dwelling house /Apartment                       NaN   
11    Second-Hand Dwelling house /Apartment                       NaN   
...                                     ...                       ...   
9876  Second-Hand Dwelling house /Apartment                       NaN   
9877  Second-Hand Dwelling house /Apartment                       NaN   
9878  Second-Hand Dwelling house /Apartment                       NaN   
9879  Second-Hand Dwelling house /Apartment                       NaN   
9873  Second-Hand Dwelling house /Apartment                       NaN   

            long        lat street_number               route  \
1     -6.1814048  53.362710           NaN  Seafield Road East   
0     -6.1814048  53.362710           NaN  Seafield Road East   
13    -6.1828326  53.362984           NaN  Seafield Road East   
12    -6.1814048  53.362710           NaN  Seafield Road East   
11    -6.1818919  53.365139             5      Seafield Grove   
...          ...        ...           ...                 ...   
9876  -6.2129274  53.376221            73    Clanmaurice Road   
9877  -6.2460341  53.294678             8     Alexandra Court   
9878  -6.1569786  53.403250             3       Dermot Street   
9879  -6.2151242  53.277574           NaN                 NaN   
9873  -6.3075642  53.390790            20    Finglaswood Road   

          neighborhood postal_town         county postal_code  
1             Clontarf    Dublin 3  County Dublin         NaN  
0             Clontarf    Dublin 3  County Dublin         NaN  
13            Clontarf    Dublin 3  County Dublin         NaN  
12            Clontarf    Dublin 3  County Dubli

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46458 entries, 1 to 9873
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Date of Sale (dd/mm/yyyy)  46458 non-null  datetime64[ns]
 1   Address                    46458 non-null  object        
 2   Postal Code                29767 non-null  object        
 3   County                     46458 non-null  object        
 4   Price (�)                  46458 non-null  float64       
 5   Not Full Market Price      46458 non-null  object        
 6   VAT Exclusive              46458 non-null  object        
 7   Description of Property    46458 non-null  object        
 8   Property Size Description  8294 non-null   object        
 9   long                       46404 non-null  object        
 10  lat                        46404 non-null  float64       
 11  street_number              33257 non-null  object        
 12  route

#### The price was read as type String, so I removed the Euro symbol that it had in the front and converted to Float

In [18]:
df[df.columns[4]] = df[df.columns[4]].replace('[\�,]', '', regex=True).astype(float)

#### Renaming some of the columns

In [20]:
df = df.rename({'Date of Sale (dd/mm/yyyy)':'Date of Sale', 'Price (�)':'Price', 'long': 'Longitude', 'lat': 'Latitude', 'neighborhood':'Neighborhood', 'postal_town': 'Postal Town'}, axis='columns')

In [21]:
df

Date of Sale                                            Address  \
1      2017-01-01  11 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
0      2017-01-01   1 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
13     2017-01-01  9 TEAL HOUSE, REDCOURT OAKS SEAFIELD RD EAST, ...   
12     2017-01-01  8 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EA...   
11     2017-01-01               5 SEAFIELD GROVE, CLONTARF, DUBLIN 3   
...           ...                                                ...   
9876   2019-08-16           73 CLANMAURICE RD, DONNYCARNEY, DUBLIN 5   
9877   2019-08-16              8 ALEXANDRA COURT, DUNDRUM, DUBLIN 16   
9878   2019-08-16          APT 1, 3 DERMOT ST, CLONGRIFFIN DUBLIN 13   
9879   2019-08-16  APT 416, THE CUBES 8, BEACON SOUTH QUARTER DUB...   
9873   2019-08-16         20 FINGLASWOOD RD, FINGLAS WEST, DUBLIN 11   

     Postal Code  County     Price Not Full Market Price VAT Exclusive  \
1       Dublin 3  Dublin  242424.0                   Yes            No   
0       Dublin 3  Dublin  242424.0                   Yes            No   
13      Dublin 3  Dublin   66000.0                   Yes            No   
12      Dublin 3  Dublin  271800.0                   Yes            No   
11      Dublin 3  Dublin  830000.0                    No            No   
...          ...     ...       ...                   ...           ...   
9876    Dublin 5  Dublin  320000.0                    No            No   
9877   Dublin 14  Dublin  285000.0                    No            No   
9878   Dublin 13  Dublin  215000.0                    No            No   
9879   Dublin 18  Dublin  365000.0                    No            No   
9873   Dublin 11  Dublin  290000.0                    No            No   

                    Description of Property Property Size Description  \
1     Second-Hand Dwelling house /Apartment                       NaN   
0     Second-Hand Dwelling house /Apartment                       NaN   
13    Second-Hand Dwelling house /Apartment                       NaN   
12    Second-Hand Dwelling house /Apartment                       NaN   
11    Second-Hand Dwelling house /Apartment                       NaN   
...                                     ...                       ...   
9876  Second-Hand Dwelling house /Apartment                       NaN   
9877  Second-Hand Dwelling house /Apartment                       NaN   
9878  Second-Hand Dwelling house /Apartment                       NaN   
9879  Second-Hand Dwelling house /Apartment                       NaN   
9873  Second-Hand Dwelling house /Apartment                       NaN   

       Longitude   Latitude street_number               route  \
1     -6.1814048  53.362710           NaN  Seafield Road East   
0     -6.1814048  53.362710           NaN  Seafield Road East   
13    -6.1828326  53.362984           NaN  Seafield Road East   
12    -6.1814048  53.362710           NaN  Seafield Road East   
11    -6.1818919  53.365139             5      Seafield Grove   
...          ...        ...           ...                 ...   
9876  -6.2129274  53.376221            73    Clanmaurice Road   
9877  -6.2460341  53.294678             8     Alexandra Court   
9878  -6.1569786  53.403250             3       Dermot Street   
9879  -6.2151242  53.277574           NaN                 NaN   
9873  -6.3075642  53.390790            20    Finglaswood Road   

          Neighborhood Postal Town         county postal_code  
1             Clontarf    Dublin 3  County Dublin         NaN  
0             Clontarf    Dublin 3  County Dublin         NaN  
13            Clontarf    Dublin 3  County Dublin         NaN  
12            Clontarf    Dublin 3  County Dublin         NaN  
11            Clontarf    Dublin 3  County Dublin    D03 K544  
...                ...         ...            ...         ...  
9876     Clontarf West    Dublin 5  County Dublin    D05 YX28  
9877  Churchtown Lower         NaN  County Dublin         NaN  
9878        Donag

#### Transforming all the empty spaces to NaN

In [22]:
import numpy as np

In [23]:
df = df.apply(lambda x: x.str.strip() if isinstance(x, str) else x).replace('', np.NaN)

In [24]:
df.head(20)

Date of Sale                                            Address  \
1    2017-01-01  11 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
0    2017-01-01   1 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
13   2017-01-01  9 TEAL HOUSE, REDCOURT OAKS SEAFIELD RD EAST, ...   
12   2017-01-01  8 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EA...   
11   2017-01-01               5 SEAFIELD GROVE, CLONTARF, DUBLIN 3   
10   2017-01-01  35 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD E...   
9    2017-01-01  32 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD E...   
2    2017-01-01                124 SEAFIELD RD, CLONTARF, DUBLIN 3   
7    2017-01-01  20 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD E...   
3    2017-01-01                126 SEAFIELD RD, CLONTARF, DUBLIN 3   
4    2017-01-01                128 SEAFIELD RD, CLONTARF, DUBLIN 3   
8    2017-01-01  25 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
6    2017-01-01  18 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
5    2017-01-01  13 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD E...   
14   2017-01-02     31 UPPER DRUMCONDRA ROAD, DRUMCONDRA, DUBLIN 9   
15   2017-01-03               114 SHELMARTIN AVE, MARINO, DUBLIN 3   
16   2017-01-03                    13 THE VIEW, ROBSWALL, MALAHIDE   
17   2017-01-03       180 THE LINKS, ELM PARK, MERRION RD DUBLIN 4   
18   2017-01-03                 4 Park lane, Belltree, Clongriffen   
19   2017-01-03  74 ST MAGDALEN TERRACE, STELLA GARDENS, IRISHT...   

   Postal Code  County     Price Not Full Market Price VAT Exclusive  \
1     Dublin 3  Dublin  242424.0                   Yes            No   
0     Dublin 3  Dublin  242424.0                   Yes            No   
13    Dublin 3  Dublin   66000.0                   Yes            No   
12    Dublin 3  Dublin  271800.0                   Yes            No   
11    Dublin 3  Dublin  830000.0                    No            No   
10    Dublin 3  Dublin  215600.0                   Yes            No   
9     Dublin 3  Dublin  352800.0                   Yes            No   
2     Dublin 3  Dublin  535500.0                   Yes            No   
7     Dublin 3  Dublin  360000.0                    No            No   
3     Dublin 3  Dublin  630000.0                    No            No   
4     Dublin 3  Dublin  535500.0                   Yes            No   
8     Dublin 3  Dublin  148148.0                   Yes            No   
6     Dublin 3  Dublin  242424.0                   Yes            No   
5     Dublin 3  Dublin  271800.0                   Yes            No   
14         NaN  Dublin  260000.0                    No            No   
15    Dublin 3  Dublin  375000.0                    No            No   
16         NaN  Dublin  575000.0                    No            No   
17    Dublin 4  Dublin  292000.0                    No            No   
18         NaN  Dublin  255507.0                    No           Yes   
19    Dublin 4  Dublin  452000.0                    No            No   

                  Description of Property  \
1   Second-Hand Dwelling house /Apartment   
0   Second-Hand Dwelling house /Apartment   
13  Second-Hand Dwelling house /Apartment   
12  Second-Hand Dwelling house /Apartment   
11  Second-Hand Dwelling house /Apartment   
10  Second-Hand Dwelling house /Apartment   
9   Second-Hand Dwelling house /Apartment   
2   Second-Hand Dwelling house /Apartment   
7   Second-Hand Dwelling house /Apartment   
3   Second-Hand Dwelling house /Apartment   
4   Second-Hand Dwelling house /Apartment   
8   Second-Hand Dwelling house /Apartment   
6   Second-Hand Dwelling house /Apartment   
5   Second-Hand Dwelling house /Apartment   
14  Second-Hand Dwelling house /Apartment   
15  Second-Hand Dwelling house /Apartment   
16  Second-Hand Dwelling house /Apartment   
17  Second-Hand Dwelling house /Apartment   
18          New Dwelling house /Apartment   
19  Second-Hand Dwelling house /Apartment   

                            Property Size Description   Longitude   Latit

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46458 entries, 1 to 9873
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Date of Sale               46458 non-null  datetime64[ns]
 1   Address                    46458 non-null  object        
 2   Postal Code                29767 non-null  object        
 3   County                     46458 non-null  object        
 4   Price                      46458 non-null  float64       
 5   Not Full Market Price      46458 non-null  object        
 6   VAT Exclusive              46458 non-null  object        
 7   Description of Property    46458 non-null  object        
 8   Property Size Description  8294 non-null   object        
 9   Longitude                  46404 non-null  object        
 10  Latitude                   46404 non-null  float64       
 11  street_number              33257 non-null  object        
 12  route

#### We can see that we have a few missing values. There are 16,691 missing values for Postal Code, 38,164 missing values for property size Description. 

#### Columns 9 to 16 were generated with the Google Maps API, because we needed to get the latitude and longitude of the addresses and because I knew we possibly would want to separate the properties by area I generated columns 11 to 16 in an attempt to get the area and neighborhood since the Postal Code column in the original dataset has a lot of missing values. It did not work so well, the only two columns that would be useful would be 'neighborhood' and 'postal-town' but unfortunately the API didn't return a result for every row, leaving us with 13,810 empty rows for neighborhood and 25,933 for postal_town. 

#### From the columns returned by the Google Maps, the only 2 that could still be usefull are 'neighborhood' and 'postal_town' so I will drop street_number, route, county and postal_code. And since we selected only Dublin data, there is no point in having a county column, because all of the entries are County Dublin

In [26]:
df['County'].unique()

array(['Dublin'], dtype=object)

In [27]:
df = df.drop(['County', 'street_number', 'route', 'county', 'postal_code'], axis =1)

In [28]:
df.shape

(46458, 12)

In [29]:
df.head()

Date of Sale                                            Address  \
1    2017-01-01  11 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
0    2017-01-01   1 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
13   2017-01-01  9 TEAL HOUSE, REDCOURT OAKS SEAFIELD RD EAST, ...   
12   2017-01-01  8 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EA...   
11   2017-01-01               5 SEAFIELD GROVE, CLONTARF, DUBLIN 3   

   Postal Code     Price Not Full Market Price VAT Exclusive  \
1     Dublin 3  242424.0                   Yes            No   
0     Dublin 3  242424.0                   Yes            No   
13    Dublin 3   66000.0                   Yes            No   
12    Dublin 3  271800.0                   Yes            No   
11    Dublin 3  830000.0                    No            No   

                  Description of Property Property Size Description  \
1   Second-Hand Dwelling house /Apartment                       NaN   
0   Second-Hand Dwelling house /Apartment                       NaN   
13  Second-Hand Dwelling house /Apartment                       NaN   
12  Second-Hand Dwelling house /Apartment                       NaN   
11  Second-Hand Dwelling house /Apartment                       NaN   

     Longitude   Latitude Neighborhood Postal Town  
1   -6.1814048  53.362710     Clontarf    Dublin 3  
0   -6.1814048  53.362710     Clontarf    Dublin 3  
13  -6.1828326  53.362984     Clontarf    Dublin 3  
12  -6.1814048  53.362710     Clontarf    Dublin 3  
11  -6.1818919  53.365139     Clontarf    Dublin 3

#### As you can see the index is messed up, it goes from 1 to 0 then 13, etc. For the iterations that we are going to do next we need to first reset the index of the dataframe

In [30]:
df = df.reset_index(drop=True)

#### Having a look at unique values for each of the columns

The column postal_town was meant to be the same as the Postal Code. It was an attempt to get the area with the Google Maps API, since there is so many missing values for Postal Code in the original dataset. Let's check what are the unique values for each

In [31]:
df['Postal Code'].unique()

array(['Dublin 3', nan, 'Dublin 4', 'Dublin 1', 'Dublin 11', 'Dublin 2',
       'Dublin 14', 'Dublin 7', 'Dublin 9', 'Dublin 10', 'Dublin 16',
       'Dublin 15', 'Dublin 6', 'Dublin 24', 'Dublin 18', 'Dublin 5',
       'Dublin 8', 'Dublin 13', 'Dublin 12', 'Dublin 20', 'Dublin 22',
       'Dublin 17', 'Dublin 6w', 'Baile �tha Cliath 18'], dtype=object)

In [32]:
df['Postal Town'].unique()

array(['Dublin 3', nan, 'Dublin 4', 'Dublin 13', 'Dublin 1', 'Dublin 11',
       'Dublin 14', 'Dublin 9', 'Dublin 10', 'Dublin 16', 'Dublin 24',
       'Dublin 6', 'Dublin 18', 'Dublin 15', 'Dublin 5', 'Dublin 8',
       'Dublin 12', 'Dublin 20', 'Dublin 22', 'Dublin 6W', 'Dublin 7',
       'Ballyboughal', 'Baile Átha Cliath 16', 'Dublin 2', 'Dublin 17',
       'Malahide', 'Dublin', 'Wolverhampton', 'Ballybrack', 'Portmarnock',
       'Castlewellan', 'Baile Átha Cliath 18', 'Dun Laoghaire',
       'Rathcoole', 'Newton Stewart', 'Rathmichael', 'Dublin, 6',
       'Baile Átha Cliath 24', 'Saggart', 'London', 'Rush', 'Westport',
       'Baile Átha Cliath 6', 'Dublin-4', 'Stillorgan', 'Barnet',
       'Blackrock', 'Swords', 'Castleknock', 'Dublin, 15', 'Skerries',
       'Calliaghstown, Rathcoole', 'Baile Átha Cliath 14', 'Phibsborough',
       'Dublin Southside', 'Newry', 'Belfast', 'Wilmslow', 'Balbriggan',
       'Newcastle upon Tyne', 'Baile Átha Cliath 9', 'Lusk', 'Donabate',
       '

#### Fixing some wrong values

In [33]:
for x in range(len(df)):
    var1 = df.loc[x, 'Postal Town']
    if (var1 == 'Baile Átha Cliath 16'):
        df['Postal Town'][x] = 'Dublin 16'
    elif (var1 == 'Baile Átha Cliath 18'):
        df['Postal Town'][x] = 'Dublin 18'
    elif (var1 == 'Baile Átha Cliath 24'):
        df['Postal Town'][x] = 'Dublin 24'
    elif ((var1 == 'Baile Átha Cliath 6') | (var1 == 'Dublin, 6') ):
        df['Postal Town'][x] = 'Dublin 6'
    elif (var1 == 'Baile Átha Cliath 14'):
        df['Postal Town'][x] = 'Dublin 14'
    elif ((var1 == 'Baile Átha Cliath 15') | (var1 == 'Dublin, 15') | (var1 == 'Dublín 15') ):
        df['Postal Town'][x] = 'Dublin 15'
    elif (var1 == 'Baile Átha Cliath 9'):
        df['Postal Town'][x] = 'Dublin 9'
    elif (var1 == 'Baile Átha Cliath 11'):
        df['Postal Town'][x] = 'Dublin 11'
    elif (var1 == 'Phibsborough'):
        df['Postal Town'][x] = 'Dublin 7'
    elif (var1 == 'Dublin-4'):
        df['Postal Town'][x] = 'Dublin 4'
    elif (var1 == 'Ballymun'):
        df['Postal Town'][x] = 'Dublin 11'
    elif ((var1 == 'Wolverhampton') | (var1 == 'Castlewellan') | (var1 == 'Newton Stewart') | (var1 == 'London') | (var1 == 'Westport') | (var1 == 'Barnet') | (var1 == 'Newry') | (var1 == 'Belfast') | (var1 == 'Wilmslow') | (var1 == 'Newcastle upon Tyne') | (var1 == 'Bushey') | (var1 == 'Glasgow') | (var1 == 'Fairbourne') | (var1 == 'Sherborne') | (var1 == 'Stirling') | (var1 == 'Cambridge') | (var1 == 'Perth') | (var1 == 'Dublin City') | (var1 == 'Dublin Southside') | (var1 == 'Dublin')):
        df['Postal Town'][x] = np.NaN #because those names are not in Dublin, the api must have found a wrong address

<ipython-input-33-c2c11b69093a>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Postal Town'][x] = 'Dublin 16'
<ipython-input-33-c2c11b69093a>:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Postal Town'][x] = np.NaN
<ipython-input-33-c2c11b69093a>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Postal Town'][x] = 'Dublin 18'
<ipython-input-33-c2c11b69093a>:10: SettingWithCopyWarning: 
A value is trying to be set on a c

In [34]:
df['Postal Town'].unique()

array(['Dublin 3', nan, 'Dublin 4', 'Dublin 13', 'Dublin 1', 'Dublin 11',
       'Dublin 14', 'Dublin 9', 'Dublin 10', 'Dublin 16', 'Dublin 24',
       'Dublin 6', 'Dublin 18', 'Dublin 15', 'Dublin 5', 'Dublin 8',
       'Dublin 12', 'Dublin 20', 'Dublin 22', 'Dublin 6W', 'Dublin 7',
       'Ballyboughal', 'Dublin 2', 'Dublin 17', 'Malahide', 'Ballybrack',
       'Portmarnock', 'Dun Laoghaire', 'Rathcoole', 'Rathmichael',
       'Saggart', 'Rush', 'Stillorgan', 'Blackrock', 'Swords',
       'Castleknock', 'Skerries', 'Calliaghstown, Rathcoole',
       'Balbriggan', 'Lusk', 'Donabate', 'Killiney', 'Newcastle',
       'Dalkey', 'Ballbriggan', 'Bray', 'Garristown'], dtype=object)

#### Investigating data types of Postal Code and Postal Town

In [35]:
var = df.loc[14, 'Postal Code']
var

nan

In [36]:
type(var)

float

In [37]:
var = df.loc[13, 'Postal Code']
var

'Dublin 3'

In [38]:
type(var)

str

In [39]:
for x in range(40):
    print(type(df.loc[x, 'Postal Town']))

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'float'>
<class 'str'>
<class 'float'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'float'>
<class 'str'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'str'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'str'>
<class 'float'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'float'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


#### We can see that apparently all the NaN is type float and the others string, what I want to do now with the loop below is check if the Postal Code (which is the original column from the PPRI) is null and the Postal Town (which was the area returned from Google Maps API) is not null, we will fill Postal Code with the value from Postal Town.

In [40]:
for x in range(len(df)):
    var1 = df.loc[x, 'Postal Code']
    var2 = df.loc[x, 'Postal Town']
    if (type(var1) == float): #if Postal Code is float means it's nan
        if (type(var2) != float): #if Postal Town is not float means it's not nan, we have a value
            df['Postal Code'][x] = df['Postal Town'][x]

<ipython-input-40-662234c85567>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Postal Code'][x] = df['Postal Town'][x]


In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46458 entries, 0 to 46457
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Date of Sale               46458 non-null  datetime64[ns]
 1   Address                    46458 non-null  object        
 2   Postal Code                32607 non-null  object        
 3   Price                      46458 non-null  float64       
 4   Not Full Market Price      46458 non-null  object        
 5   VAT Exclusive              46458 non-null  object        
 6   Description of Property    46458 non-null  object        
 7   Property Size Description  8294 non-null   object        
 8   Longitude                  46404 non-null  object        
 9   Latitude                   46404 non-null  float64       
 10  Neighborhood               32648 non-null  object        
 11  Postal Town                20323 non-null  object        
dtypes: d

#### Before we had 29767 values for Postal Code and now we have 32607. We could get 2,840 more values for the Area/Postal Code but we are still missing 13,851. And we can now drop Postal Town since it is redundant

In [42]:
df = df.drop(['Postal Town'], axis =1)

In [43]:
df

Date of Sale                                            Address  \
0       2017-01-01  11 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
1       2017-01-01   1 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
2       2017-01-01  9 TEAL HOUSE, REDCOURT OAKS SEAFIELD RD EAST, ...   
3       2017-01-01  8 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EA...   
4       2017-01-01               5 SEAFIELD GROVE, CLONTARF, DUBLIN 3   
...            ...                                                ...   
46453   2019-08-16           73 CLANMAURICE RD, DONNYCARNEY, DUBLIN 5   
46454   2019-08-16              8 ALEXANDRA COURT, DUNDRUM, DUBLIN 16   
46455   2019-08-16          APT 1, 3 DERMOT ST, CLONGRIFFIN DUBLIN 13   
46456   2019-08-16  APT 416, THE CUBES 8, BEACON SOUTH QUARTER DUB...   
46457   2019-08-16         20 FINGLASWOOD RD, FINGLAS WEST, DUBLIN 11   

      Postal Code     Price Not Full Market Price VAT Exclusive  \
0        Dublin 3  242424.0                   Yes            No   
1        Dublin 3  242424.0                   Yes            No   
2        Dublin 3   66000.0                   Yes            No   
3        Dublin 3  271800.0                   Yes            No   
4        Dublin 3  830000.0                    No            No   
...           ...       ...                   ...           ...   
46453    Dublin 5  320000.0                    No            No   
46454   Dublin 14  285000.0                    No            No   
46455   Dublin 13  215000.0                    No            No   
46456   Dublin 18  365000.0                    No            No   
46457   Dublin 11  290000.0                    No            No   

                     Description of Property Property Size Description  \
0      Second-Hand Dwelling house /Apartment                       NaN   
1      Second-Hand Dwelling house /Apartment                       NaN   
2      Second-Hand Dwelling house /Apartment                       NaN   
3      Second-Hand Dwelling house /Apartment                       NaN   
4      Second-Hand Dwelling house /Apartment                       NaN   
...                                      ...                       ...   
46453  Second-Hand Dwelling house /Apartment                       NaN   
46454  Second-Hand Dwelling house /Apartment                       NaN   
46455  Second-Hand Dwelling house /Apartment                       NaN   
46456  Second-Hand Dwelling house /Apartment                       NaN   
46457  Second-Hand Dwelling house /Apartment                       NaN   

        Longitude   Latitude      Neighborhood  
0      -6.1814048  53.362710          Clontarf  
1      -6.1814048  53.362710          Clontarf  
2      -6.1828326  53.362984          Clontarf  
3      -6.1814048  53.362710          Clontarf  
4      -6.1818919  53.365139          Clontarf  
...           ...        ...               ...  
46453  -6.2129274  53.376221     Clontarf West  
46454  -6.2460341  53.294678  Churchtown Lower  
46455  -6.1569786  53.403250        Donaghmede  
46456  -6.2151242  53.277574        Blackthorn  
46457  -6.3075642  53.390790               NaN  

[46458 rows x 11 columns]

Exploring the values for some of the features

In [44]:
df['Not Full Market Price'].unique()

array(['Yes', 'No'], dtype=object)

In [45]:
df['VAT Exclusive'].unique()

array(['No', 'Yes'], dtype=object)

In [46]:
df['Description of Property'].unique()

array(['Second-Hand Dwelling house /Apartment',
       'New Dwelling house /Apartment', 'Teach/�ras�n C�naithe Ath�imhe'],
      dtype=object)

In [47]:
df['Property Size Description'].unique()

array([nan,
       'greater than or equal to 38 sq metres and less than 125 sq metres',
       'greater than or equal to 125 sq metres', 'less than 38 sq metres'],
      dtype=object)

#### We can see that Property Size Description was considered as: 
* less than 38 sq metres
* greater than or equal to 38 sq metres and less than 125 sq metres
* greater than or equal to 125 sq metres
#### It would be a very useful feature but we have only 8294 values out of 46458 so aprox. 82% of the values are missing, because of that I am going to drop that column as well

In [48]:
df = df.drop(['Property Size Description'], axis =1)

In [49]:
df

Date of Sale                                            Address  \
0       2017-01-01  11 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
1       2017-01-01   1 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EAST   
2       2017-01-01  9 TEAL HOUSE, REDCOURT OAKS SEAFIELD RD EAST, ...   
3       2017-01-01  8 PINTAIL HOUSE, REDCOURT OAKS, SEAFIELD RD EA...   
4       2017-01-01               5 SEAFIELD GROVE, CLONTARF, DUBLIN 3   
...            ...                                                ...   
46453   2019-08-16           73 CLANMAURICE RD, DONNYCARNEY, DUBLIN 5   
46454   2019-08-16              8 ALEXANDRA COURT, DUNDRUM, DUBLIN 16   
46455   2019-08-16          APT 1, 3 DERMOT ST, CLONGRIFFIN DUBLIN 13   
46456   2019-08-16  APT 416, THE CUBES 8, BEACON SOUTH QUARTER DUB...   
46457   2019-08-16         20 FINGLASWOOD RD, FINGLAS WEST, DUBLIN 11   

      Postal Code     Price Not Full Market Price VAT Exclusive  \
0        Dublin 3  242424.0                   Yes            No   
1        Dublin 3  242424.0                   Yes            No   
2        Dublin 3   66000.0                   Yes            No   
3        Dublin 3  271800.0                   Yes            No   
4        Dublin 3  830000.0                    No            No   
...           ...       ...                   ...           ...   
46453    Dublin 5  320000.0                    No            No   
46454   Dublin 14  285000.0                    No            No   
46455   Dublin 13  215000.0                    No            No   
46456   Dublin 18  365000.0                    No            No   
46457   Dublin 11  290000.0                    No            No   

                     Description of Property   Longitude   Latitude  \
0      Second-Hand Dwelling house /Apartment  -6.1814048  53.362710   
1      Second-Hand Dwelling house /Apartment  -6.1814048  53.362710   
2      Second-Hand Dwelling house /Apartment  -6.1828326  53.362984   
3      Second-Hand Dwelling house /Apartment  -6.1814048  53.362710   
4      Second-Hand Dwelling house /Apartment  -6.1818919  53.365139   
...                                      ...         ...        ...   
46453  Second-Hand Dwelling house /Apartment  -6.2129274  53.376221   
46454  Second-Hand Dwelling house /Apartment  -6.2460341  53.294678   
46455  Second-Hand Dwelling house /Apartment  -6.1569786  53.403250   
46456  Second-Hand Dwelling house /Apartment  -6.2151242  53.277574   
46457  Second-Hand Dwelling house /Apartment  -6.3075642  53.390790   

           Neighborhood  
0              Clontarf  
1              Clontarf  
2              Clontarf  
3              Clontarf  
4              Clontarf  
...                 ...  
46453     Clontarf West  
46454  Churchtown Lower  
46455        Donaghmede  
46456        Blackthorn  
46457               NaN  

[46458 rows x 10 columns]

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46458 entries, 0 to 46457
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date of Sale             46458 non-null  datetime64[ns]
 1   Address                  46458 non-null  object        
 2   Postal Code              32607 non-null  object        
 3   Price                    46458 non-null  float64       
 4   Not Full Market Price    46458 non-null  object        
 5   VAT Exclusive            46458 non-null  object        
 6   Description of Property  46458 non-null  object        
 7   Longitude                46404 non-null  object        
 8   Latitude                 46404 non-null  float64       
 9   Neighborhood             32648 non-null  object        
dtypes: datetime64[ns](1), float64(2), object(7)
memory usage: 3.5+ MB


#### Apparently Longitude is as String and Latitude as Float. This was because one or more entries had a comma in front of the number, for instance: ',-6.23454', that I was my fault when manually entering a few numbers, since there was a few addresses the api didn't find, I googled the address and manually entered the latitude and longitude, so I probably copied with a comma. That should be numeric so I will delete the comma and convert it to Float.

In [51]:
df[df.columns[7]] = df[df.columns[7]].replace('[\�,]', '', regex=True).astype(float)

#### There are a few missing values for Latitude and Longitude, these are values that I couldn't understand what the correct address was, and I prefered not to input it since I could be putting the wrong place. So I am just going to drop those ones

In [53]:
df = df.loc[df['Longitude'].notnull()]

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46404 entries, 0 to 46457
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date of Sale             46404 non-null  datetime64[ns]
 1   Address                  46404 non-null  object        
 2   Postal Code              32573 non-null  object        
 3   Price                    46404 non-null  float64       
 4   Not Full Market Price    46404 non-null  object        
 5   VAT Exclusive            46404 non-null  object        
 6   Description of Property  46404 non-null  object        
 7   Longitude                46404 non-null  float64       
 8   Latitude                 46404 non-null  float64       
 9   Neighborhood             32646 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(6)
memory usage: 3.9+ MB


#### Just checking if indexes need to be updated again

In [55]:
df.tail()

Date of Sale                                            Address  \
46453   2019-08-16           73 CLANMAURICE RD, DONNYCARNEY, DUBLIN 5   
46454   2019-08-16              8 ALEXANDRA COURT, DUNDRUM, DUBLIN 16   
46455   2019-08-16          APT 1, 3 DERMOT ST, CLONGRIFFIN DUBLIN 13   
46456   2019-08-16  APT 416, THE CUBES 8, BEACON SOUTH QUARTER DUB...   
46457   2019-08-16         20 FINGLASWOOD RD, FINGLAS WEST, DUBLIN 11   

      Postal Code     Price Not Full Market Price VAT Exclusive  \
46453    Dublin 5  320000.0                    No            No   
46454   Dublin 14  285000.0                    No            No   
46455   Dublin 13  215000.0                    No            No   
46456   Dublin 18  365000.0                    No            No   
46457   Dublin 11  290000.0                    No            No   

                     Description of Property  Longitude   Latitude  \
46453  Second-Hand Dwelling house /Apartment  -6.212927  53.376221   
46454  Second-Hand Dwelling house /Apartment  -6.246034  53.294678   
46455  Second-Hand Dwelling house /Apartment  -6.156979  53.403250   
46456  Second-Hand Dwelling house /Apartment  -6.215124  53.277574   
46457  Second-Hand Dwelling house /Apartment  -6.307564  53.390790   

           Neighborhood  
46453     Clontarf West  
46454  Churchtown Lower  
46455        Donaghmede  
46456        Blackthorn  
46457               NaN

In [56]:
df = df.reset_index(drop=True)

In [57]:
df.tail()

Date of Sale                                            Address  \
46399   2019-08-16           73 CLANMAURICE RD, DONNYCARNEY, DUBLIN 5   
46400   2019-08-16              8 ALEXANDRA COURT, DUNDRUM, DUBLIN 16   
46401   2019-08-16          APT 1, 3 DERMOT ST, CLONGRIFFIN DUBLIN 13   
46402   2019-08-16  APT 416, THE CUBES 8, BEACON SOUTH QUARTER DUB...   
46403   2019-08-16         20 FINGLASWOOD RD, FINGLAS WEST, DUBLIN 11   

      Postal Code     Price Not Full Market Price VAT Exclusive  \
46399    Dublin 5  320000.0                    No            No   
46400   Dublin 14  285000.0                    No            No   
46401   Dublin 13  215000.0                    No            No   
46402   Dublin 18  365000.0                    No            No   
46403   Dublin 11  290000.0                    No            No   

                     Description of Property  Longitude   Latitude  \
46399  Second-Hand Dwelling house /Apartment  -6.212927  53.376221   
46400  Second-Hand Dwelling house /Apartment  -6.246034  53.294678   
46401  Second-Hand Dwelling house /Apartment  -6.156979  53.403250   
46402  Second-Hand Dwelling house /Apartment  -6.215124  53.277574   
46403  Second-Hand Dwelling house /Apartment  -6.307564  53.390790   

           Neighborhood  
46399     Clontarf West  
46400  Churchtown Lower  
46401        Donaghmede  
46402        Blackthorn  
46403               NaN

#### Writing to csv

In [58]:
df.to_csv('df_allyears_coords.csv')